packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -0,0 +1,250 @@
+          ...config.redis.subhub,
+        },
+        log
+      );
+    this.plansCacheIsEnabled = this.cacheTtlSeconds && redis;
+
+    this.stripe = new stripe(config.subscriptions.stripeApiKey, {
+      maxNetworkRetries: 3,
+    });
+    this.redis = redis;
+  }
+
+  /**
+   * Fetch all product data and cache it if Redis is enabled.
+   *
+   * Use `allProducts` below to use the cached-enhanced version.
+   *
+   * @returns {Promise<AbbrevProduct[]>} All the products.
+   */
+  async fetchAllProducts() {
+    const products = [];
+    for await (const product of this.stripe.products.list()) {
+      products.push({
+        product_id: product.id,
+        product_name: product.name,
+        product_metadata: product.metadata,
+      });
+    }
+    return products;
+  }
+
+  /**
+   * Fetches all products from stripe and returns them.
+   *
+   * Uses Redis caching if configured.
+   *
+   * @returns {Promise<AbbrevProduct[]>} All the products.
+   */
+  async allProducts() {
+    return cachedResult(
+      this.log,
+      this.redis,
+      'listProducts',
+      this.cacheTtlSeconds,
+      () => this.fetchAllProducts()
+    );
+  }
+
+  /**
+   * Fetches all plans from stripe and returns them.
+   *
+   * Use `allPlans` below to use the cached-enhanced version.
+   *
+   * @returns {Promise<AbbrevPlan[]>} All the plans.
+   */
+  async fetchAllPlans() {
+    const plans = [];
+    for await (const item of this.stripe.plans.list({
+      active: true,
+      expand: ['data.product'],
+    })) {
+      // FIXME: Should probably error here if we can't set a product id/name.
+      let product_id, product_name, product_metadata;
+      if (item.product && typeof item.product !== 'string') {
+        product_id = item.product.id;
+        product_name = item.product.name;
+        product_metadata = item.product.metadata;
+      }
+
+      plans.push({
+        plan_id: item.id,
+        plan_name: item.nickname,
+        plan_metadata: item.metadata,
+        product_id,
+        product_name,
+        product_metadata,
+        interval: item.interval,
+        amount: item.amount,
+        currency: item.currency,
+      });
+    }
+    return plans;
+  }
+
+  /**
+   * Fetches all plans from stripe and returns them.
+   *
+   * Uses Redis caching if configured.
+   *
+   * @returns {Promise<AbbrevPlan[]>} All the plans.
+   */
+  async allPlans() {
+    return cachedResult(
+      this.log,
+      this.redis,
+      'listPlans',
+      this.cacheTtlSeconds,
+      () => this.fetchAllPlans()
+    );
+  }
+
+  /**
+   * Verify that the `planId` is a valid upgrade for this `productId`.
+   *
+   * Throws an error if its an invalid upgrade.
+   *
+   * @param {string} productId
+   * @param {string} planId
+   * @returns {Promise<void>}
+   */
+  async verifyPlanUpgradeForSubscription(productId, planId) {
+    const allPlans = await this.allPlans();
+    const currentPlan = allPlans
+      .filter(plan => plan.product_id === productId)
```

Comment:
Instead of `filter().shift()`, you can just use [`find`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/find) I think?